### PR TITLE
Introduce TsuAsyncGeterr for prevent blocking Vim's UI

### LINKED ISSUE #241: File saving takes multiple seconds
Hi,

I'm running into issues where saving a .ts file will cause vim to freeze for a few seconds. It doesn't always happen, but it happens more often than not. 

vim version
`VIM - Vi IMproved 8.0 (2016 Sep 12, compiled Mar 13 2018 21:01:05)`

profile.log
```
FUNCTIONS SORTED ON TOTAL TIME
count  total (s)   self (s)  function
    2   7.603904   0.000071  tsuquyomi#reloadAndGeterr()
    2   7.603768   0.000104  tsuquyomi#geterr()
    2   7.603664   0.000381  tsuquyomi#createFixlist()
    2   7.602379   0.000133  tsuquyomi#tsClient#tsGeterr()
    2   7.602218   0.000472  tsuquyomi#tsClient#sendCommandSyncEvents()
    2   7.599539   7.599446  tsuquyomi#tsClient#sendRequest()
   38   0.043134   0.001774  <SNR>138_VimOutputCallback()
   38   0.039921   0.000871  <SNR>139_HandleCommandMessage()
   38   0.038612   0.003276  ale#lsp#HandleMessage()
   48   0.026647   0.001670  ale#engine#HandleLSPResponse()
   18   0.026341   0.000855  ale#CallWithCooldown()
   30   0.024977   0.015433  <SNR>96_HandleTSServerDiagnostics()
   11   0.019284             <SNR>75_Highlight_Matching_Pair()
    2   0.012150   0.000167  ale#events#SaveEvent()
    2   0.011819   0.000069  ale#Queue()
    2   0.011625   0.000231  <SNR>63_ALEQueueImpl()
    2   0.009969   0.000058  ale#Lint()
    2   0.009782   0.000285  <SNR>63_ALELintImpl()
    6   0.009380   0.000301  ale#engine#HandleLoclist()
    2   0.008206   0.000522  ale#engine#RunLinters()

FUNCTIONS SORTED ON SELF TIME
count  total (s)   self (s)  function
    2   7.599539   7.599446  tsuquyomi#tsClient#sendRequest()
   11              0.019284  <SNR>75_Highlight_Matching_Pair()
   30   0.024977   0.015433  <SNR>96_HandleTSServerDiagnostics()
   38              0.007078  ale#lsp#ReadMessageData()
    2              0.006193  <SNR>2_StripTrailingWhitespaces()
    2              0.004118  prettier#Autoformat()
   38   0.038612   0.003276  ale#lsp#HandleMessage()
  109              0.002901  ale#Var()
    6              0.002614  ale#path#FindNearestFile()
   34   0.002645   0.002126  <SNR>27_repo_head_ref()
   68              0.001881  <SNR>27_repo()
   24   0.003746   0.001786  ale#ShouldDoNothing()
   38   0.043134   0.001774  <SNR>138_VimOutputCallback()
   48   0.026647   0.001670  ale#engine#HandleLSPResponse()
   11   0.002360   0.001605  <SNR>93_OnCursorMovedNormalMode()
   34   0.005983   0.001547  <SNR>27_repo_head()
    6              0.001415  ale#sign#ParseSigns()
    4   0.002000   0.001298  ale#linter#Get()
   48              0.001198  ale#lsp#HandleOtherInitializeResponses()
   38              0.001060  ale#job#ParseVim8ProcessID()

```

this line in the function seems to be the culript:
`   30              7.594757       let out = ch_readraw(s:tsq['channel'])`

Any advice? Thanks.

### PR DESCRIPTION
**Pull request by heavenshell**
Closes #241 #250  #274 

Currently `tsuGeterr` sometimes slow.
- Responses of TSServer is slow
- `tsuquyomi#tsClient#sendRequest` wait TSServer's responses synchronous

I introduce async feature to tsuquyomi and this change will prevent blocking Vim.

- Add TsuAsyncGeterr

### Before

Start Vim and run `:TsuGeterr` block Vim's UI.

### After

Start Vim and run `:TsuAsyncGeterr` does not block Vim's UI.
After run `:TsuAsyncGeterr`, you can operate Vim(Not blocking).

### How to work

Set `let g:tsuquyomi_use_vimproc = 0` to your `.vimrc`.
This PR use Vim8's job and channel feature.

Execute `:tsuAsyncGeterr`

or add followings to your `.vimrc`.
```
autocmd InsertLeave,TextChanged,BufWritePost *.ts,*.tsx call tsuquyomi#asyncGeterr()
```

### TODO
After this PR would accepted, I'll add `event queue` mechanism to debounce `send change event to TSServer`. 
`TextChanged` would send buffer to TSServer when text are chaned.
We don't need to send buffer each time.

Also I try to add test, but in my local, did not ran test well.
It would be glad to add `tsAsyncGeterr.spec.vim`.

### CODE REVIEW

File: autoload/tsuquyomi.vim [hunk 1]
```diff
@@ -556,37 +571,85 @@ endfunction
...
+    endfor
+  endif
+  return quickfix_list
+endfunction
+
+function! tsuquyomi#registerNotify(callback, eventName)
+  call tsuquyomi#tsClient#registerNotify(a:callback, a:eventName)
+endfunction
+
+function! tsuquyomi#emitChange(bufnum)
```
Comment (heavenshell) [inline autoload/tsuquyomi.vim]: `s:flush()` is slow.
Just send `change` to TSServer.

File: autoload/tsuquyomi.vim [hunk 2]
```diff
@@ -556,37 +571,85 @@ endfunction
...
+
+function! tsuquyomi#asyncCreateFixlist(...)
+  " Works only Vim8(+channel, +job)
+  " We must register callbacks(handler and callback) before execute this.
+  " See `tsuquyomi#config#initBuffer()`
+  if len(s:checkOpenAndMessage([expand('%:p')])[1])
+    return []
+  endif
+  " `tsuquyomi#getSupportedCodeFixes()` is too slow and block Vim's ui.
+  " call tsuquyomi#getSupportedCodeFixes()
```
Comment (heavenshell) [inline autoload/tsuquyomi.vim]: `tsuquyomi#getSupportedCodeFixes()` is also slow.
So I remove this at async.

File: autoload/tsuquyomi.vim [hunk 3]
```diff
@@ -556,37 +571,85 @@ endfunction
...
+endfunction
+
 function! tsuquyomi#createQuickFixListFromEvents(event_list)
   if !len(a:event_list)
     return []
   endif
   let quickfix_list = []
   let supportedCodes = tsuquyomi#getSupportedCodeFixes()
   for event_item in a:event_list
-    if has_key(event_item, 'type') && event_item.type ==# 'event' && (event_item.event ==# 'syntaxDiag' || event_item.event ==# 'semanticDiag')
```
Comment (heavenshell) [inline autoload/tsuquyomi.vim]: Move to `tsuquyomi#parseDiagnosticEvent()` for commonize.

File: autoload/tsuquyomi.vim [hunk 4]
```diff
@@ -94,6 +94,21 @@ function! s:writeToPreview(content)
   setlocal nomodifiable readonly
   silent wincmd p
 endfunction
+
+function! s:setqflist(quickfix_list, ...)
+  " 0: Do not close cwindow automatically
+  " 1: Close cwindow automatically
+  let auto_close = len(a:000) ? a:0 : 0
```
Comment (heavenshell) [inline autoload/tsuquyomi.vim]: Notify errors to user should do silent.
If Hooking TsuAsyncGeterr to Texchanged autocmd, Open / Close cwindow many times.
This is annoying.

**Comment by Quramy** [comment]
Thanks @heavenshell 
I’ll review and try this :)

File: autoload/tsuquyomi/tsClient.vim [hunk 5]
```diff
@@ -29,14 +29,19 @@ endif
 
 let s:request_seq = 0
 
-let s:ignore_respons_conditions = []
+let s:ignore_response_conditions = []
```
[the remark this answers is not in the document]
Comment (Quramy) [reply autoload/tsuquyomi/tsClient.vim]: oops 😭

**Comment by Quramy** [comment]
@heavenshell I've read your changes. It's so great ! 😄 

I add some nits comment, please check them 🙏

File: autoload/tsuquyomi/tsClient.vim [hunk 6]
```diff
@@ -156,10 +175,77 @@ function! tsuquyomi#tsClient#statusTss()
...
+  let l:eventName = s:getEventType(l:item)
+
+  if(has_key(s:callbacks, l:eventName))
+    let Callback = function(s:callbacks[l:eventName], [l:item])
+    call Callback()
+  endif
+endfunction
+
+function! tsuquyomi#tsClient#clearCallbacks()
+  let s:callback_list = {}
```
[the remark this answers is not in the document]
Comment (heavenshell) [reply autoload/tsuquyomi/tsClient.vim]: That's right. Thank you. Fixed.

**Comment by heavenshell** [comment]
Thank you for review and pointing out.
Fixed. Please take another look 🙇

**Comment by Quramy** [comment]
Thanks for your great contribution!

**Comment by artemave** [comment]
This is great! Any plans for `asyncGeterrProject`?

**Comment by DGmip** [comment]
Thank you so much!!

**Pull request merged**